Exposes Baggage Apis

adriancole opened the pull request:
This completes the most common requests around propagated fields, both synchronizing
to logging context and also baggage propagation.

Notably, this allows you to not only define constants for the fields you
want, but also decide if they are permissible for correlation.

Ex.

```java
 // configure the field, permitting its value to be immediately flushed to a logging context
 static final BaggageField BUSINESS_PROCESS = BaggageField.newBuilder("bp").flushOnUpdate().build();

 // Allow logging patterns like %X{traceId} %X{bp}
 decorator = MDCScopeDecorator.newBuilder()
                              .addField(BUSINESS_PROCESS).build();

 tracingBuilder.propagationFactory(BaggagePropagation.newFactoryBuilder(B3Propagation.FACTORY)
                                                     .addRemoteField(BUSINESS_PROCESS)
                                                     .build())
               .currentTraceContext(ThreadLocalCurrentTraceContext.newBuilder()
                                                                  .addScopeDecorator(decorator)
                                                                  .build())

 // Assume runtime code looks like this
 @SendTo(SourceChannels.OUTPUT)
 public void timerMessageSource() {
   BUSINESS_PROCESS.setValue("accounting");
   // The logging expression %{bp} will show "accounting" in businessCode()
   businessCode();
 }
```

## W3C baggage header
W3C has a header encoding that could be used for baggage, but it is unstable
https://github.com/w3c/correlation-context/issues/17

A future version can be compatible with w3c baggage, which allows multiple in-process fields to be encoded into the same remote field. For now, it will throw an exception.

Ex.
```java
BaggagePropagation.FactoryBuilder builder = newFactoryBuilder(B3Propagation.FACTORY)
     .addRemoteField(BaggageField.create("userName", baggage")
     .addRemoteField(BaggageField.create("userId", baggage"); // throws IllegalArgumentException
```

## Spring integration

The baggage fields types are made to integrate cleanly with Spring. For example, `BaggageField` is an input to `CorrelationScopeDecorator` and `BaggagePropagation`, but only `BaggagePropagation` controls whether or not it is propagated remotely.

Fixes #682

adriancole commented:
the code needs a little work, but I wanted to raise this for feedback as the brunt of it is done.

adriancole commented on brave/src/main/java/brave/propagation/ExtraField.java:
```diff
@@ -0,0 +1,330 @@
+import static brave.propagation.ExtraFieldPropagation.currentTraceContext;
+
+/**
+ * Defines a request-scoped field, usually but not always analogous to an HTTP header. Fields will
```
@jorgheymans  can you check this new type to see if it makes sense?

anuraaga commented on brave/src/main/java/brave/propagation/ExtraField.java:
```diff
@@ -0,0 +1,331 @@
+  }
+
+  /** Extracts the value for this field in the context or null if unavailable. */
+  @Nullable public String getValue(TraceContext context) {
```
I guess these can delegate to the static methods instead of copy-paste

adriancole replied:
this drifted after we now use the field type internally.

anuraaga commented on brave/src/test/java/brave/features/opentracing/OpenTracingAdapterTest.java:
```diff
@@ -37,12 +36,12 @@
   List<zipkin2.Span> spans = new ArrayList<>();
   Tracing brave = Tracing.newBuilder()
-    .currentTraceContext(ThreadLocalCurrentTraceContext.newBuilder()
-      .addScopeDecorator(StrictScopeDecorator.create())
```
Think you may have accidentally lost the strict checker

adriancole replied:
not needed as not multi-threaded.. sorry I was cleaning up.

anuraaga commented on context/log4j2/src/main/java/brave/context/log4j2/Log4j2Context.java:
```diff
@@ -0,0 +1,32 @@
+import org.apache.logging.log4j.ThreadContext;
+
+/** Integrates Log4j 2's {@link org.apache.logging.log4j.ThreadContext} */
+public final class Log4j2Context extends CorrelationField.Context {
```
Since package name is `log4j2` think this can just be `Log4jContext`

adriancole replied:
I was going by the marketing on their website and kinda prefer the match on package name..

adriancole replied:
obviated as the type is hidden

adriancole commented on brave/src/main/java/brave/propagation/CorrelationField.java:
```diff
@@ -0,0 +1,48 @@
+  /** Dispatches methods to synchronize fields with a context such as SLF4J MDC */
+  abstract class Context { // abstract to allow optimized methods to be added later
+    /**
+     * Returns the correlation property of the specified name iff it is a string, or null
```
some contexts allow object values, so the main intent here is that it is not null and a string..

adriancole commented on brave/src/main/java/brave/propagation/CorrelationField.java:
```diff
@@ -0,0 +1,48 @@
+  @Nullable String getValue(TraceContext context);
+
+  /** Dispatches methods to synchronize fields with a context such as SLF4J MDC */
+  abstract class Context { // abstract to allow optimized methods to be added later
```
oh yeah I remember that. I didn't really want this to be a user type though. In fact, we can probably hide it again..

adriancole replied:
reason is that the context is very expensive part of the code, and having to re-look up the old value will just make it worse.

adriancole commented on brave/src/main/java/brave/propagation/CorrelationFields.java:
```diff
@@ -0,0 +1,127 @@
+
+    @Override public boolean equals(Object o) {
+      if (o == this) return true;
+      if (!(o instanceof BaseCorrelationField)) return false;
```
sure.

adriancole replied:
I will try to seal it, handling extra field differently..

adriancole commented on brave/src/main/java/brave/propagation/CorrelationFields.java:
```diff
@@ -0,0 +1,127 @@
+   *
+   * <p>Ex.
+   * <pre>{@code
+   * awsRegion = CorrelationFields.constant("region", System.getEnv("AWS_REGION"));
```
heh I actually looked up a real variable :D fair enough though.

adriancole commented on brave-tests/src/main/java/brave/test/ITRemote.java:
```diff
@@ -79,7 +80,8 @@ protected TraceContext newTraceContext(SamplingFlags flags) {
 
   protected final Propagation.Factory propagationFactory =
-    ExtraFieldPropagation.newFactory(B3Propagation.FACTORY, EXTRA_KEY);
+    ExtraFieldPropagation.newFactoryBuilder(B3Propagation.FACTORY)
```
thanks for the thought although I don't really want to couple propagation to this type yet..

right now there's a generic list<extra> thing in the context also, so I don't want to conflate the two.

adriancole commented on brave/src/main/java/brave/internal/PredefinedPropagationFields.java:
```diff
@@ -78,12 +78,12 @@ public String get(int index) {
-    if (index >= fieldNames.length) return;
+    if (index >= names.length) return;
 
     synchronized (this) {
```
drifted to ExtraFields but will comment now!

putAllIfAbsent is protected and used before the input is returned anywhere. I'll add a javadoc comment. Regardless I made a path that doesn't synchronize even if it would be uncontended. Good idea.

adriancole commented on brave/src/main/java/brave/propagation/CorrelationField.java:
```diff
@@ -0,0 +1,48 @@
+import brave.internal.Nullable;
+
+/**
+ * Defines a field handling in a correlation context such as MDC.
```
better thanks

adriancole commented on brave/src/main/java/brave/propagation/CorrelationField.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Defines a field handling in a correlation context such as MDC.
+ *
+ * <p>{@link #equals(Object)} and {@link #hashCode()} should be overridden to implement value
```
I reworded ptal. can't be sealed as we need it to be an interface for conditional use in ExtraField..

adriancole commented on brave/src/main/java/brave/propagation/CorrelationFieldScopeDecorator.java:
```diff
@@ -0,0 +1,219 @@
+          scope.close();
+          replace(field.name(), previousValue);
+        }
+      }
```
made a comment elsewhere, but to read back the current value is expensive. also we allow late updates. finally we do want to clobber out of band updates because MDC can be used externally

adriancole commented on brave/src/main/java/brave/propagation/CorrelationField.java:
```diff
@@ -0,0 +1,48 @@
+  String name();
+
+  /** Extracts the value of the field named {@link #name()} in the context or null if unavailable. */
+  @Nullable String getValue(TraceContext context);
```
I'd rather use returns because it shouldn't be expensive like extract is. correlation field is in-process. updated.

adriancole commented on brave/src/main/java/brave/propagation/CorrelationFields.java:
```diff
@@ -0,0 +1,127 @@
+    }
+  };
+  /**
+   * This is only useful when {@link #SPAN_ID} is also a correlation field.
```
span id isn't a request id, but anyway would like to dodge a discussion :)

trying "Used with {@link #TRACE_ID} to correlate a log line with a span."

adriancole commented:
@anuraaga thanks for all the feedback I think it is better now. The code has drifted since your last view as it now uses ExtraField type internally.

Next step for me is to make the flush integration work. The idea is to use a thread local for the scope made by CorrelationFieldScopeDecorator (this will include some refactoring for the two variants that exist there). ExtraField.setValue() will look at that thread-local and act accordingly.

As mentioned a few times, `CorrelationField.Context` isn't a user type. We can probably make it internal or hidden or loudly mentioned don't use this. It is only used to dispatch to MDC which users can muck with (or other libraries can) anyway. I don't really want to do more work in this type about previous vs current value comparison as the largest amount of overhead is actually the MDC impls.

adriancole commented:
actually what you mention is good for another reason.. we don't have to verify the prior value if we are passed it. However, we can't verify it if we are not. @anuraaga  I think I like your idea for a different reason. Right now, we have StrictScopeDecorator which checks some things. While it is probably a bad idea to do any more use of MDC than we have to, we could make a strict Correlation context impl that someone can use when testing instrumentation. The notes about performance in the other cases can then move into their respective impls, and shortcut where costly. With the more defensive type signature, we can be a little more permissive with it.. for example, warning less loudly. Thanks for the idea!

adriancole commented:
looking at how things work i'm less inclined to expose the context type. Rather, I think we can hide it by fusing Log4J2Context etc into a subtype of CorrelationFieldScopeDecorator.Builder for now. Injecting/auto-wiring builders is fairly commonplace.

adriancole commented:
unconditionally reverting fields on scope close will still probably imply too much overhead. Saving off "last values" is also high overhead as it means array allocations. However, there's another way out. If we restrict ExtraFields to no more than 32 fields, we can use an integer bitflag for a dirty bit. If there's a late update, we set a bit and revert that one. Otherwise it is effectively a noop. I'll implement this tomorrow.

adriancole commented:
also by 32 fields I mean 32 not including aliases like one field prefixed two ways. ex the field "foo" prefixed as "baggage-foo" and "baggage_foo" (ExtraField.keys()) is still one field.

jorgheymans commented on brave/src/main/java/brave/propagation/ExtraField.java:
```diff
@@ -0,0 +1,330 @@
+ * Defines a request-scoped field, usually but not always analogous to an HTTP header. Fields will
+ * be no-op unless {@link ExtraFieldPropagation} is configured.
+ *
+ * <p>For example, if you have a need to know the a specific request's country code, you can
```
remove 'the'

jorgheymans commented on brave/src/main/java/brave/propagation/ExtraField.java:
```diff
@@ -0,0 +1,330 @@
+ * spanCustomizer.tag(countryCode.name(), countryCode.getValue(context));
+ * spanCustomizer.tag("country-code", ExtraField.getValue("country-code"));
+ *
+ * // You can also set or override the value similarly, which might be needed if a new request
```
'for a new request' rather ? otherwise the sentence looked incomplete at first sight - though grammatically correct

adriancole replied:
You can also update the value similarly, so that the new value will propagate downstream.

comment back if needed

jorgheymans commented on brave/src/main/java/brave/propagation/ExtraField.java:
```diff
@@ -0,0 +1,330 @@
+ *
+ * <p>Brave is an infrastructure library: you will create lock-in if you expose its apis into
+ * business code. Prefer exposing your own types for utility functions that use this class as this
+ * will insulate you from lock-in.
```
this is already mentioned in brave's main README.md , any merit repeating the obvious here ? The security bits are very relevant but this applies to all public types in brave no ?

adriancole replied:
The reason this is in the type doc is that this type specifically adds temptation to embed in your code, and readme doesn't inherit docs and visa versa (maybe asciidoc can help if someone can help with it)

That said, this doc was ignored on the type it was on last time, leading to this very change, heh!

jorgheymans replied:
gotcha.

jorgheymans commented on brave/src/main/java/brave/propagation/ExtraField.java:
```diff
@@ -0,0 +1,330 @@
+ * business code. Prefer exposing your own types for utility functions that use this class as this
+ * will insulate you from lock-in.
+ *
+ * <p>While it may seem convenient, do not use this for security context propagation as it was not
```
if the paragraph about the infrastructure library gets removed consider merging this one and the `it is generally not a good idea ...` into one.

jorgheymans commented on brave/src/main/java/brave/propagation/ExtraField.java:
```diff
@@ -0,0 +1,330 @@
+  }
+
+  /**
+   * Sets the value of the field with the specified key, or drops if not a configured field.
```
`or drops if not a configured field` what do you mean it's just a no-op call like ignored ? `or ignored if the key does not reference a configured field` perhaps

adriancole replied:
thx

jorgheymans commented on brave/src/main/java/brave/propagation/ExtraField.java:
```diff
@@ -0,0 +1,330 @@
+  }
+
+  /**
+   * Sets the current value of the field with the specified key, or drops if not a configured
```
same comment as above

jorgheymans commented on brave/src/main/java/brave/propagation/ExtraField.java:
```diff
@@ -0,0 +1,330 @@
+    return PropagationFields.get(context, name, ExtraFieldPropagation.Extra.class);
+  }
+
+  /** Sets the value of the this field, or drops if not configured. */
```
same `drops` comment as above

jorgheymans commented on brave/src/main/java/brave/propagation/ExtraField.java:
```diff
@@ -0,0 +1,330 @@
+  }
+
+  /**
+   * Sets the value of this field in the implicit trace context or ignores when not available.
```
here `ignores` is being used, wording is more clear to me

adriancole commented:
after a not restful sleep I realize the flushing feature is expensive especially as it subverts the `maybeScope` feature we wrote for reactive code (cc @kojilin). When we have insecurity about a possible late update, we are forced to track fields (allocate overhead) even if nothing is ever set.

To accomodate this, I will change the contract a bit. "Updatable" subinterface of CorrelationField, defined as dangerous for performance, and is only used to allow an update in-between newScope and close. Those who choose this will have to pay for it, but others will not.

adriancole commented:
by classifying fields that are allowed to update mid-scope, we can make things cheap enough routinely to save performance and introduce the thing @anuraaga discussed.

Also, we can be like grpc and not fail if the last value isn't current.

adriancole commented:
ps Flushable isn't the same as Updatable. field updates don't impact performance unless we are expected to write them immediately to the underlying context. Ex field updates apply now at the next scope operation.

adriancole commented:
flushable probably won't describe well enough the nuance.

the field needs to have a change policy attribute
ex. span, scope, immediate

for example, currentTraceContext.maybeScope returns Scope.NOOP if the trace/span ID is the same. Since initial fields are read prior to creating the context, they are visible and would be assumed to apply formerly if you receive Scope.NOOP.

Extra fields set late can apply at the next scope operation, but only if Scope.NOOP is re-evaluated. For example, we can look at the current value of the fields to see if they expect a change, then revert that later. This is expensive as it requires looking at the logging context for each scope operation, but not as expensive as the last.

The most expensive is flushing as we have to defensively save off values even if they are the same at the entrance of the scope, and revert them when the scope is closed. flushing means Scope.NOOP can never be returned.

adriancole commented:
will rebase after https://github.com/openzipkin/brave/pull/1131

anuraaga commented on brave/src/main/java/brave/propagation/ExtraFields.java:
```diff
@@ -0,0 +1,180 @@
+    return -1;
+  }
+
+  @Override public int hashCode() { // for unit tests
```
Even only for unit tests, probably good to compare `fields` too or if a newly added test doesn't work because of that in the future it will be quite mysterious

adriancole replied:
went offline for a while as I really don't want to do this.. I'm rejigging the tests instead.

adriancole replied:
glad I didn't want to change the hashCode/equals.. there's a bug in the impl :)

anuraaga commented on context/log4j12/README.md:
```diff
@@ -1,14 +1,15 @@
     .currentTraceContext(ThreadLocalCurrentTraceContext.newBuilder()
-       .addScopeDecorator(MDCScopeDecorator.create())
+        .addScopeDecorator(CorrelationFieldScopeDecorator.create(correlationContext))
        .build()
```
```suggestion
        .build()
```

anuraaga commented on context/log4j2/README.md:
```diff
@@ -2,13 +2,14 @@
     .currentTraceContext(ThreadLocalCurrentTraceContext.newBuilder()
-       .addScopeDecorator(ThreadContextScopeDecorator.create())
+        .addScopeDecorator(CorrelationFieldScopeDecorator.create(correlationContext))
        .build()
```
```suggestion
        .build()
```

anuraaga commented on context/slf4j/README.md:
```diff
@@ -2,13 +2,14 @@
     .currentTraceContext(ThreadLocalCurrentTraceContext.newBuilder()
-       .addScopeDecorator(MDCScopeDecorator.create())
+        .addScopeDecorator(CorrelationFieldScopeDecorator.create(correlationContext))
        .build()
```
```suggestion
        .build()
```

anuraaga commented on brave/src/main/java/brave/propagation/ExtraField.java:
```diff
@@ -0,0 +1,409 @@
+ * <h3>Prefixed fields</h3>
+ * <p>You can also prefix fields, if they follow a common pattern. For example, the following will
+ * propagate the field "x-vcap-request-id" as-is, but send the fields "country-code" and "user-id"
+ * on the wire as "baggage-country-code" and "baggage-user-id" respectively.
```
Realized that baggage + country code is fun too

adriancole replied:
these are actual header names prefixes. ex sleuth adds the baggage- prefix. you'll find a lot where we intentionally use real examples so that things aren't so abstract. This isn't just test json in zipkin, rather all the projects.

Is there a prefix you've mapped that would be better I tried to find a header prefixed userinfo and couldn't..

adriancole commented on brave/src/main/java/brave/propagation/CorrelationField.java:
```diff
@@ -0,0 +1,61 @@
+  /**
+   * Marks a field such that updates immediately flush to the correlation context, when configured
+   * and their corresponding trace context is current. This has a significant performance impact as
+   * it will invalidate the {@link CurrentTraceContext#maybeScope(TraceContext)} operator.
```
I rewrote as this:

```
     * Call this to immediately flush a value update to the correlation context as opposed waiting
     * for the next scope decoration. This has a significant performance impact as it requires even
     * {@link CurrentTraceContext#maybeScope(TraceContext)} to always track values.
```

maybeScope returns no-op when applying the context results in the same data. maybeScope when not interfered with prevents another layer (ex several arrays). If we don't know if a field will be flushed or not, we need to allocate all of that even if no change ever takes place. This could happen thousands of times.

adriancole commented on brave/src/main/java/brave/propagation/CorrelationFieldScopeDecorator.java:
```diff
@@ -0,0 +1,243 @@
+ * // Any scope operations (updates to the current span) apply the fields defined by the decorator.
+ * ScopedSpan span = tracing.tracer().startScopedSpan("encode");
+ * try {
+ *   // %X{traceId} %X{parentId} %X{spanId} %X{sampled} %X{region} are in the logging context!
```
good idea

adriancole commented on brave/src/main/java/brave/propagation/CorrelationFieldScopeDecorator.java:
```diff
@@ -0,0 +1,243 @@
+  // This means we cannot return a no-op scope based on if we detect no change when comparing
+  // values up front. Hence, we save off the first value and revert when a scope closes. If a late
+  // update changed the value mid-scope, it will reverted.
+  void updateUnconditionally(CorrelationField field, @Nullable String newValue) {
```
already changed to update :)

adriancole commented on brave/src/main/java/brave/propagation/CorrelationFieldScopeDecorator.java:
```diff
@@ -0,0 +1,243 @@
+  public static abstract class Builder {
+    final CorrelationContext context;
+    final Set<CorrelationField> fields = new LinkedHashSet<>(asList(
+      CorrelationFields.TRACE_ID,
```
I suppose we can for the builder variant. we still need all fields for the backported factory method. Sounds good!

adriancole commented on brave/src/main/java/brave/propagation/CorrelationFields.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    @Override public String name() {
+      return name;
```
no this is case sensitive.. it is only the extra keys (not name) that are downcase. 

existing correlation patterns use 'traceId' not 'traceid' for example.

I put a TODO to seal this type.

adriancole commented on brave/src/main/java/brave/propagation/ExtraField.java:
```diff
@@ -0,0 +1,409 @@
+     *
+     * @since 5.11
+     */
+    public CorrelationBuilder withCorrelation() {
```
actually redacted was made exactly for logging context (exists now and is in use in sleuth)

the docs say "Sets this field to be only visible in process, redacted from remote propagation."

adriancole commented on brave/src/main/java/brave/propagation/ExtraField.java:
```diff
@@ -0,0 +1,409 @@
+    return PropagationFields.get(context, this, ExtraFields.class);
+  }
+
+  /** Updates the value of the this field, or ignores if not configured. */
```
ExtraFieldPropagation is the component that configures extra fields and that's the first line of the type. comment later if you have another thing we have to do on this.

adriancole commented on brave/src/main/java/brave/propagation/ExtraField.java:
```diff
@@ -0,0 +1,409 @@
+  /** @since 5.11 */
+  public static class Builder {
+    final String name;
+    final Set<String> keys = new LinkedHashSet<>();
```
key is already used in Propagation component, that's what these names are literally for. I added `@see ExtraFieldPropagation#keys()` to the thing.

adriancole commented on brave/src/main/java/brave/propagation/ExtraFieldPropagation.java:
```diff
@@ -15,246 +15,197 @@
+        String name = entry.getKey();
+        ExtraField.Builder builder = ExtraField.newBuilder(name);
+        if (redactedNames.contains(name)) builder.redacted();
+        if (!remainingNames.remove(name)) builder.clearKeys();
```
I'm overloaded so I'll get back to this..

adriancole commented on brave/src/main/java/brave/propagation/ExtraFields.java:
```diff
@@ -0,0 +1,180 @@
+
+    MapFieldConsumer result = new MapFieldConsumer();
+    forEach(result);
+    return result;
```
saved also for tomorrow.

adriancole commented:
as far as I know the code probably works now, but I have to write the tests on it to make sure. Until this happens, please don't give feedback on javadoc as it will drift. I also am not planning to do any more polishing on things that weren't polished prior to this PR as it is taking a massive amount of time away to chase all these things down.

IOTW, let's please hold off more workload until this works, and then please only work load related to this change that must be done here and can't be done after by someone else.

adriancole commented:
Just renamed everything to baggage per https://github.com/w3c/correlation-context/issues/17

adriancole commented:
OK thanks for your patience everyone. Please do review the code at this point with the exception of `CorrelationScopeDecorator` internals and test cases, which I'm finishing up. If you have Karma I'd appreciate just pushing any grammatical or consistency related changes directly to this branch instead of having me broker them.

Thanks for the support!

adriancole commented:
PS I will revert the deleted types for ExtraFieldPropagation.. don't worry :) ExtraFieldPropagation will become a subtype of BaggagePropagation

adriancole commented on brave/src/main/java/brave/internal/MapPropagationFields.java:
```diff
@@ -59,18 +59,18 @@ protected MapPropagationFields(MapPropagationFields<K, V> parent) {
         // this is the copy-on-write part
         values = new LinkedHashMap<>(values);
-        values.put(key, value);
+        if (value != null) values.put(key, value);
```
we just newed up values so we don't need an else right?

anuraaga replied:
We newed by copying `values`. If `value` is null, don't we need to remove it from the copy?

adriancole replied:
you were right! we were missing a test

adriancole commented on brave/src/main/java/brave/propagation/BaggagePropagation.java:
```diff
@@ -0,0 +1,262 @@
+    }
+
+    /** Returns a wrapper of the delegate if there are no fields to propagate. */
+    public Factory build() {
```
we need to expose baggageKeys() for opentracing..

anuraaga replied:
Ah got it

adriancole replied:
wait maybe we don't anymore .. lemme check!

adriancole replied:
yeah we do.. I'll make a note

adriancole replied:
actually I think we can maybe cheat. lemme try..

adriancole replied:
wow that was nice! cheating works fine

adriancole commented on brave/src/main/java/brave/propagation/BaggagePropagation.java:
```diff
@@ -0,0 +1,262 @@
+
+    void inject(PredefinedBaggageFields fields, C carrier) {
+      for (BaggageFieldWithKeys<K> fieldWithKeys : propagation.fieldsWithKeys) {
+        if (fieldWithKeys.field.local) continue; // don't propagate downstream
```
yep good idea!

adriancole commented on brave/src/main/java/brave/propagation/BaggageField.java:
```diff
@@ -0,0 +1,296 @@
+     *
+     * @since 5.11
+     */
+    public Builder local() {
```
yep sounds good!

adriancole commented:
finished test backfilling for the core types

TODO:
* revert the deleted types for ExtraFieldPropagation
  * ExtraFieldPropagation will become a subtype of BaggagePropagation
* clean up internals of CorrelationScopeDecorator
* backfill tets for CorrelationField and CorrelationFields

adriancole commented:
blocked on https://github.com/openzipkin/brave/pull/1133

adriancole commented:
force pushed with only change since last night as backporting the old extra fields code.

TODO:
 * rebase over the Tags thing and start integrating that
 * clean up internals of CorrelationScopeDecorator
 * backfill tets for CorrelationField and CorrelationFields

adriancole commented:
added this to the description

## W3C baggage header
W3C has a header encoding that could be used for baggage, but it is unstable
https://github.com/w3c/correlation-context/issues/17

A future version can be compatible with w3c baggage, which allows multiple in-process fields to be encoded into the same remote field. For now, it will throw an exception.

Ex.
```java
BaggagePropagation.FactoryBuilder builder = newFactoryBuilder(B3Propagation.FACTORY);    builder.addField(BaggageField.newBuilder("userName").addRemoteName("baggage").build());
builder.addField(BaggageField.newBuilder("userId").addRemoteName("baggage").build());
builder.build(); // will throw UnsupportedOperationException
```

adriancole commented:
I've re-organized the code. @jorgheymans @anuraaga while I go back and backfill tests, please feel free to reword things or polish code by pushing updates to this branch. Chat or raise a comment if there's something you aren't sure about. It would be very appreciated!

adriancole commented:
scope decorator code is now 100% coverage. moving on to correlation field (also need to make the Tag variant for it)

adriancole commented:
collapsed the types into one: `BaggageField` correlation field is no more.

adriancole commented:
only class left for coverage is `BaggageFields`

any last feedback welcome otherwise will merge tomorrow morning after testing with sleuth

adriancole commented:
one last change is how "remote names" are configured. now that correlation field and baggage field are the same type, we can consider baggage the same as traceID. We don't change the trace context in order to say how it is serialized. similarly we shouldn't change a baggage field to do that. Once this is the case, not only is a lot of ancillary config hidden, but also we can be more flexible in inject/extract supporting exactly the same features as before, including downstream redaction.

adriancole commented:
OK final version done.. will be testing in sleuth now

adriancole commented:
opentracing.. https://github.com/openzipkin-contrib/brave-opentracing/pull/101

next to try sleuth without making any changes except using both of these versions

adriancole commented:
going to merge on green. Thanks for epic help over the last week @anuraaga!

@jorgheymans thanks for offering to tidy up more post merge!
